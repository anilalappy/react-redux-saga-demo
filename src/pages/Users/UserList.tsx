import { useEffect, useState } from "react";
import {
  TableContainer,
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableRow,
} from "@material-ui/core";
import { getRequest } from "../../utils/services";
import { insertionSort } from "../../utils/algorithms";

const createData = (
  name: string,
  calories: number,
  fat: number,
  carbs: number,
  protein: number
) => {
  return { name, calories, fat, carbs, protein };
};

const rows = [
  createData("Frozen yoghurt", 159, 6.0, 24, 4.0),
  createData("Ice cream sandwich", 237, 9.0, 37, 4.3),
  createData("Eclair", 262, 16.0, 24, 6.0),
  createData("Cupcake", 305, 3.7, 67, 4.3),
  createData("Gingerbread", 356, 16.0, 49, 3.9),
];

const UserList = () => {
  const [users, setUsers] = useState<any>([]);
  // insertionSort([5, 2, 4, 6, 1, 3]);
  const getUserList = async () => {
    const users = await getRequest("users");
    setUsers(users);
  };
  useEffect(() => {
    getUserList();
  }, []);

  return (
    <TableContainer>
      <Table
        style={{ width: "100%", justifyContent: "center", textAlign: "center" }}
        aria-label="simple table"
      >
        <TableHead>
          <TableRow>
            <TableCell>ID</TableCell>
            <TableCell>Name</TableCell>
            <TableCell>Username</TableCell>
            <TableCell>Email</TableCell>
            <TableCell>City</TableCell>
            <TableCell>Zipcode</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {users.map(({ id, name, username, email, address }: any) => (
            <TableRow
              key={id}
              // sx={{ '&:last-child td, &:last-child th': { border: 0 } }}
            >
              <TableCell component="th" scope="row">
                {id}
              </TableCell>
              <TableCell>{name}</TableCell>
              <TableCell>{username}</TableCell>
              <TableCell>{email}</TableCell>
              <TableCell>{address?.city}</TableCell>
              <TableCell>{address?.zipcode}</TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  );
};
export default UserList;
