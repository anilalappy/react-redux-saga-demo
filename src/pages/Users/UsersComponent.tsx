import { Container, Typography } from "@material-ui/core";
import { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";
import CardContainer from "../../components/common/CardContainer";
import { NoData } from "../../components/common/Container";
import Loader from "../../components/common/Loader";
import { getUsers } from "../../redux/actions/users";
import { User } from "../../redux/types";

const Users = () => {
  const dispatch = useDispatch();
  const users: User[] = useSelector((state: any) => state?.users?.users);
  const loading: boolean = useSelector((state: any) => state?.users?.loading);
  const error: string = useSelector((state: any) => state?.users?.error);
  const actionMessage: any = useSelector((state: any) => state?.users);
  if(users && users?.length > 0){
    users.sort(function(a, b) {
      return ((b.id as number) - (a.id as number));
  })
  }
  useEffect(() => {
    if (actionMessage.users && actionMessage.users.length < 1)
      dispatch(getUsers([]));
  }, []);

  return (
    <Container data-testid="user-container">
      {loading && <Loader />}
      <div
        style={{ textDecoration: "none", textAlign: "right", right: "50px" }}
      >
        <Link
          data-testid="add-user-link"
          to={`/add-edit`}
          style={{ textDecoration: "none", color: "#3f51b5" }}
        >
          Add New User
        </Link>
      </div>
      <hr />
      {!loading &&
        users &&
        users?.length > 0 &&
        users.map((user: User) => <CardContainer user={user} key={user?.id} />)}

      {error && !loading && <Typography>{error}</Typography>}
    </Container>
  );
};
export default Users;
