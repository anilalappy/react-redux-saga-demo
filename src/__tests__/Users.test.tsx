import Users from "../pages/Users/UsersComponent";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import { render } from "@testing-library/react";

const mockStore = configureStore([]);

const initialState = {
  users: [
    {
      id: 1,
      name: "Leanne Graham",
      username: "Bret",
      email: "Sincere@april.biz",
      address: {
        city: "Gwenborough",
        zipcode: "92998-3874",
      },
    },
  ],
  loading: false,
  error: null,
};

// for snapshot testing we used renderer method
describe("verify user container with snap shot", () => {
  let store: any;
  let wrapper: any;

  beforeEach(() => {
    store = mockStore({ users: initialState });
    store.dispatch = jest.fn();
    wrapper = renderer.create(
      <BrowserRouter>
        <Provider store={store}>
          <Users />
        </Provider>
      </BrowserRouter>
    );
  });
  it("user component snapshot verification", () => {
    // arrange
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

});
describe("verify user container", () => {
  let store: any;
  beforeEach(() => {
    store = mockStore({ users: initialState });
    store.dispatch = jest.fn();
  });
  it("render user component without error", () => {
    const { getByTestId } = render(
      <BrowserRouter>
        <Provider store={store}>
          <Users />
        </Provider>
      </BrowserRouter>
    );
    expect(getByTestId("user-container")).toBeInTheDocument();
    expect(getByTestId("add-user-link")).toBeInTheDocument();
  });
});
