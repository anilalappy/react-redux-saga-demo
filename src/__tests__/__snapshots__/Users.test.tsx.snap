// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verify user container with snap shot user component snapshot verification 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthLg"
  data-testid="user-container"
>
  <div
    style={
      Object {
        "right": "50px",
        "textAlign": "right",
        "textDecoration": "none",
      }
    }
  >
    <a
      data-testid="add-user-link"
      href="/add-edit"
      onClick={[Function]}
      style={
        Object {
          "color": "#3f51b5",
          "textDecoration": "none",
        }
      }
    >
      Add New User
    </a>
  </div>
  <hr />
  <div
    style={
      Object {
        "backgroundColor": "#bcbcf7",
        "border": "1px solid #6161ed",
        "borderRadius": "10px",
        "margin": "1rem 0 1rem 0",
        "width": "100%",
      }
    }
  >
    <div
      className="MuiGrid-root MuiGrid-container"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-9"
      >
        <div
          className="MuiCardContent-root"
        >
          <p
            className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
            style={
              Object {
                "fontSize": 14,
              }
            }
          >
            Name: 
            Leanne Graham
          </p>
          <p
            className="MuiTypography-root MuiTypography-body1"
            style={
              Object {
                "marginBottom": 1.5,
              }
            }
          >
            Username: 
            Bret
          </p>
          <h6
            className="MuiTypography-root MuiTypography-subtitle2"
            style={
              Object {
                "marginBottom": 1.5,
              }
            }
          >
            Email: 
            Sincere@april.biz
          </h6>
          <h6
            className="MuiTypography-root MuiTypography-subtitle2"
            style={
              Object {
                "marginBottom": 1.5,
              }
            }
          >
            Address: 
          </h6>
          <h6
            className="MuiTypography-root MuiTypography-subtitle2"
            style={
              Object {
                "marginBottom": 1.5,
                "marginLeft": "0.5rem",
              }
            }
          >
            City: 
            Gwenborough
          </h6>
          <h6
            className="MuiTypography-root MuiTypography-subtitle2"
            style={
              Object {
                "marginBottom": 1.5,
                "marginLeft": "0.5rem",
              }
            }
          >
            ZipCode: 
            92998-3874
          </h6>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
        style={
          Object {
            "margin": "auto",
            "textAlign": "left",
          }
        }
      >
        <a
          href="/add-edit/1"
          onClick={[Function]}
          style={
            Object {
              "textDecoration": "none",
            }
          }
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall"
            disabled={false}
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              Edit
            </span>
          </button>
        </a>
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSecondary MuiButton-textSizeSmall MuiButton-sizeSmall"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Delete
          </span>
        </button>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1"
      />
    </div>
  </div>
</div>
`;
